Window to worker communication series (#4340)
GoogleChrome/web.dev: src/site/content/en/reliable/broadcast-updates-guide/index.md

tags:
  - service-worker
  - offline
---

In some scenarios the service worker might need to proactively communicate with any of the active
tabs it controls to inform of a certain event. Examples include:

- Informing the page when a new version of the service worker has been installed, so that the page
  can show an **"Update to refresh"** button to the user to access the new functionality
  immediately.
- Letting the user know about a change on cached data that took place on the service worker side, by
  showing an indication, like: **"The app is now ready to work offline"**, or **"New version of the
  content available"**.

<figure class="w-figure">
  <img src="broadcast-updates-diagram.png"
       width="550"
       alt="Diagram showing a service worker communicating with the page to send an update.">
</figure>

We'll call these types of use cases where the service worker doesn't need to receive a message from
the page to start a communication **“broadcast updates”**. In this guide we'll review different
ways of implementing this type of communication between pages and service workers, by using standard

> browser APIs and the [Workbox library](https://developers.google.com/web/tools/workbox).
> 
> ## Production cases {: #production-cases }

Reviewer: ```suggestion
browser APIs and the [Workbox library](https://developers.google.com/web/tools/workbox).

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the
  [Window to workers communication](/reliable/#window-to-worker-communication) series
  for guides on other common use cases.
{% endAside %}

## Production cases {: #production-cases }
```

</figure>

### Squoosh {: #squoosh }

In the [Squoosh PWA](https://squoosh.app/), when the service worker has cached all of the necessary
assets to make it work offline, it sends a message to the page to show a “Ready to work offline”
toast, letting the user know about the feature:

<figure class="w-figure">
  <img src="squoosh-screenshot.png"
       width="550"
       alt="A screenshot of Squoosh webapp 'Ready to work offline' functionality.">
  <figcaption class="w-figcaption">In the Squoosh PWA the service worker broadcasts an update to the page when cache is ready, and the page displays “Ready to work offline” toast.
</figcaption>
</figure>

## Using Workbox {: #using-workbox }

### Listen to service worker lifecycle events {: #listen-to-service-worker-lifecycle-events }

`workbox-window` provides a straightforward interface to listen to [important service worker lifecycle
events](https://developers.google.com/web/tools/workbox/modules/workbox-window#important_service_worker_lifecycle_moments).
Under the hood, the library uses client-side APIs like
[`updatefound`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/onupdatefound)

> and [vstatechange`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/onstatechange)

Reviewer: ```suggestion
and [statechange](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/onstatechange)
```

---
layout: post 
title: "Broadcast Updates" 
authors:
  - demianrenzulli
  - andrewguan 
date: 2020-12-04 
description: | 
  How service workers can proactively communicate with the page to inform about certain events.
tags:
  - blog
  - service-worker

>   - web-worker

Author: Done

---

In some scenarios the service worker might need to proactively communicate with any of the active
tabs it controls to inform of a certain event. Examples include:

- Informing the page when a new version of the service worker has been installed, so that the page
  can show an **"Update to refresh"** button to the user to access the new functionality
  immediately.
- Letting the user know about a change on cached data that took place on the service worker side, by
  showing an indication, like: **"The app is now ready to work offline"**, or **"New version of the
  content available"**.

<figure class="w-figure">
  <img src="broadcast-updates-diagram.png"
       width="550"
       alt="Diagram showing a service worker communicating with the page to send an update.">
</figure>

We'll call these types of use cases where the service worker doesn't need to receive a message from

> the page to start a communication **“broadcast updates”**. In this article we'll review different

Reviewer: ```suggestion
the page to start a communication **“broadcast updates”**. In this guide we'll review different
```
https://documentation.divio.com/how-to-guides/

ways of implementing this type of communication between pages and service workers, by using standard
browser APIs and the [Workbox library](https://developers.google.com/web/tools/workbox).

## Production cases

### Tinder

Tinder PWA uses [Workbox

> Window](https://developers.google.com/web/tools/workbox/modules/workbox-window) to listen to

Reviewer: See imperative caching for reminder on how to format `workbox-window`, plz fix throughout the doc

Author: Done

important service worker lifecycle moments from the page (“installed”, “controlled” and
“activated”). That way when a new service worker comes into play, it shows an **"Update Available"**
banner, so that they can refresh the PWA and access the latest features:

<figure class="w-figure">
  <img src="tinder-screenshot.png"
       width="650"
       alt="A screenshot of Tinder's webapp 'Update Available' functionality.">
  <figcaption class="w-figcaption">In the Tinder PWA, the service worker tells the page that a new version is ready, and the page shows users a “Update Available” banner.</figcaption>
</figure>

### Squoosh

In the [Squoosh PWA](https://squoosh.app/), when the service worker has cached all of the necessary
assets to make it work offline, it sends a message to the page to show a “Ready to work offline”
toast, letting the user know about the feature:

<figure class="w-figure">
  <img src="squoosh-screenshot.png"
       width="550"
       alt="A screenshot of Squoosh webapp 'Ready to work offline' functionality.">

>   <figcaption class="w-figcaption">In the Sqoosh PWA the service worker broadcasts an update to the page when cache is ready, and the page displays “Ready to work offline” toast.

Reviewer: Squoosh

Author: Done

</figcaption>
</figure>

## Using Workbox

### Listen to service worker lifecycle events

Workbox Window provides a straightforward interface to listen to [important service worker lifecycle
events](https://developers.google.com/web/tools/workbox/modules/workbox-window#important_service_worker_lifecycle_moments).
Under the hood, the library uses client-side APIs like

> [onupdatefound](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/onupdatefound)

Reviewer: Reminder about formatting event names

Author: Done

To broadcast updates add a `broadcastUpdate.BroadcastUpdatePlugin` to your strategy options in the
service worker side:

```javascript
import {registerRoute} from 'workbox-routing';
import {StaleWhileRevalidate} from 'workbox-strategies';
import {BroadcastUpdatePlugin} from 'workbox-broadcast-update';

registerRoute(
  ({url}) => url.pathname.startsWith('/api/'),
  new StaleWhileRevalidate({
    plugins: [
      new BroadcastUpdatePlugin(),
    ],
  })
);
```

In your web app, you can listen for these events like so:

```javascript
navigator.serviceWorker.addEventListener('message', async (event) => {
  // Optional: ensure the message came from workbox-broadcast-update
  if (event.data.meta === 'workbox-broadcast-update') {
    const {cacheName, updatedUrl} = event.data.payload;

    // Do something with cacheName and updatedUrl.
    // For example, get the cached content and update
    // the content on the page.
    const cache = await caches.open(cacheName);
    const updatedResponse = await cache.match(updatedUrl);
    const updatedText = await updatedResponse.text();
  }
});
```

> ## Using Browser APIs

Reviewer: ```suggestion
## Using browser APIs
```
Plz double-check that all your section headings are sentence case, not title case

If the functionality that Workbox provides is not enough for your needs, use the following browser
APIs to implement **“broadcast updates”**:

### Broadcast Channel API

The service worker creates a [BroadcastChannel
object](https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel) and starts sending
messages to it. Any context (e.g. page) interested in receiving these messages can instantiate a
`BroadcastChannel` object and implement a message handler to receive messages.

To inform the page when a new service worker is installed, use the following code:

```javascript

> //Create Broadcast Channel to send messages to the page

Reviewer: No space between the comment characters `//` and the comment message?

Author: Fixed this

  if (clients && clients.length) {
    //Respond to last focused tab
    clients[0].postMessage({type: 'MSG_ID'});
  }
});
```

The page implements a message handler to intercept these messages:

```javascript
//listen to messages
navigator.serviceWorker.onmessage = (event) => {
     if (event.data && event.data.type === 'MSG_ID') { 
         //process response
   } 
};
```

Client API is a great option for cases like broadcasting information to multiple active tabs. The
API is supported by all major browsers, but not all of its methods are. Check browser support before
using it.

### Message Channel

[Message Channel](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API) requires
an initial configuration step, by passing a port from the page to the service worker, to establish a
communication channel between them. The page instantiates a `MessageChannel` object and passes a
port to the service worker, via the `postMessage()` interface:

```javascript
const messageChannel = new MessageChannel();

//Init port
navigator.serviceWorker.controller.postMessage({type: 'PORT_INITIALIZATION'}, [
  messageChannel.port2,
]);
```

> The page listens to message by implementing an “onmessage” handler on that port:

Reviewer: ```suggestion
The page listens to messages by implementing an “onmessage” handler on that port:
```
Reminder about formatting message names: `message` not "onmessage"

```javascript
//Listen to messages
messageChannel.port1.onmessage = (event) => {
  // Process message
};
```

The service worker receives the port and saves a reference to it:

```javascript
//Initialize
let communicationPort;

self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'PORT_INITIALIZATION') {
    communicationPort = event.ports[0];
  }
});
```

From that point it can send messages to the page, by calling `postMessage()` in the reference to the
port:

```javascript
//Communicate
communicationPort.postMessage({type: 'MSG_ID' });
```

`MessageChannel` might be more complex to implement, due to the need of initializing ports, but it's
supported by [all major browsers](https://caniuse.com/?search=channel%20messaging).

## Next steps

In this guide we explored one particular case of Window to service worker communication:
**"broadcast updates"**. The examples explored include listening to important service worker

> lifecycle events, and communicating the page about changes in content or cached data. You can think

Reviewer: ```suggestion
lifecycle events, and communicating to the page about changes in content or cached data. You can think
```

- Informing the page when a new version of the service worker has been installed, so that the page
  can show an **"Update to refresh"** button to the user to access the new functionality
  immediately.
- Letting the user know about a change on cached data that took place on the service worker side, by
  showing an indication, like: **"The app is now ready to work offline"**, or **"New version of the
  content available"**.

<figure class="w-figure">
  <img src="broadcast-updates-diagram.png"
       width="550"
       alt="Diagram showing a service worker communicating with the page to send an update.">
</figure>

We'll call these types of use cases where the service worker doesn't need to receive a message from
the page to start a communication **“broadcast updates”**. In this guide we'll review different
ways of implementing this type of communication between pages and service workers, by using standard
browser APIs and the [Workbox library](https://developers.google.com/web/tools/workbox).

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the

>   [Window to workers communication](/reliable/#window-to-worker-communication) series

Reviewer: ```suggestion
  [Window to workers communication](/reliable/#communicate-with-workers) series
```
